tree/dataframe/src/RDFSnapshotHelpers.cxx

```diff
@@ -930,3 +995,260 @@ ROOT::Internal::RDF::UntypedSnapshotRNTupleHelper::MakeNew(void *newName)
+         }
+         return;
+      }
+
+      // Neither branch nor systematic are known, so a new entry needs to be created
+      fNBits = std::max(fNBits, variationIndex);
+      const auto vectorIndex = variationIndex / 64u;
+      const auto bitIndex = variationIndex % 64u;
+
+      // Create bitmask branches as long as necessary to capture the bit
+      while (vectorIndex >= fBitMasks.size()) {
+         std::string bitmaskBranchName =
+            std::string{"R_rdf_mask_"} + fTree->GetName() + '_' + std::to_string(fBitMasks.size());
+         fBitMasks.push_back(Bitmask{bitmaskBranchName});
+         fTree->Branch(bitmaskBranchName.c_str(), fBitMasks.back().branchBuffer.get());
+      }
+
+      fBranchToVariationMapping[branchName] = variationIndex;
+      fBranchToBitmaskMapping[branchName] = std::make_pair(fBitMasks[vectorIndex].branchName, bitIndex);
+   }
+
+   /// Clear all bits, as if none of the variations passed its filter.
+   void ClearMaskBits()
+   {
+      for (auto &mask : fBitMasks)
+         mask.bitset.reset();
+   }
+
+   /// Set a bit signalling that the variation at `index` passed its filter.
+   void SetMaskBit(unsigned int index)
+   {
+      const auto vectorIndex = index / 64;
+      const auto bitIndex = index % 64;
+      fBitMasks[vectorIndex].bitset.set(bitIndex, true);
+   }
+
+   /// Test if any of the mask bits are set.
+   bool MaskEmpty() const
+   {
+      return std::none_of(fBitMasks.begin(), fBitMasks.end(), [](Bitmask const &mask) { return mask.bitset.any(); });
+   }
+
+   /// Write the current event and the bitmask to the output dataset.
+   void Write() const
+   {
+      if (!fTree)
+         throw std::runtime_error("The TTree associated to the Snapshot action doesn't exist, any more.");
+
+      for (auto const &mask : fBitMasks) {
+         *mask.branchBuffer = mask.bitset.to_ullong();
+      }
+
+      fTree->Fill();
+   }
+};
+
+ROOT::Internal::RDF::SnapshotHelperWithVariations::SnapshotHelperWithVariations(
+   std::string_view filename, std::string_view dirname, std::string_view treename, const ColumnNames_t &vbnames,
+   const ColumnNames_t &bnames, const RSnapshotOptions &options, std::vector<bool> &&isDefine,
+   ROOT::Detail::RDF::RLoopManager *outputLoopMgr, ROOT::Detail::RDF::RLoopManager *inputLoopMgr,
+   const std::vector<const std::type_info *> &colTypeIDs)
+   : fOptions(options), fInputLoopManager{inputLoopMgr}, fOutputLoopManager{outputLoopMgr}
+{
+   EnsureValidSnapshotTTreeOutput(fOptions, std::string(treename), std::string(filename));
+
+   TFile::TContext fileCtxt;
+   fOutputHandle = std::make_shared<SnapshotOutputWriter>(
+      TFile::Open(filename.data(), fOptions.fMode.c_str(), /*ftitle=*/"",
+                  ROOT::CompressionSettings(fOptions.fCompressionAlgorithm, fOptions.fCompressionLevel)),
+      nullptr);
+   if (!fOutputHandle->fFile)
+      throw std::runtime_error(std::string{"Snapshot: could not create output file "} + std::string{filename});
```

Comment:
See my other comment in the constructor of `SnapshotOutputWriter`

Reply:
I added the suggested assertion, but for release builds, we will need to keep this exception. Furthermore, the OutputWriter doesn't know the filename if file creation fails, so the exception needs to be where we open the file.